feat: release ADK-alpha (#1)
tier4/perception_ecu_container: README.md

# perception\_ecu\_container

This repository provides a sample environment working with Jetson + GMSL cam + ROS2.

The following figure shows a system overview.

> ![](docs/overview.svg "system overview")

Reviewer: [style] I believe `markdownlint` doesn't allow empty image names.

```suggestion
![system overview](docs/overview.svg "system overview")
```

Author: https://github.com/tier4/perception_ecu_container/pull/1/commits/fe9a4ad2f69e2f2e76dc354da67043cd13319a2d

#### [for RQX-58G from ADLINK Technology, Inc. users] BSP installation
This repository assumes RQX-58Gs are correctly set up by following to the official quick start guide from ADLINK Technology, Inc. Please see [the official document](https://www.adlinktech.com/Products/Download.ashx?type=MDownload&isQuickStart=yes&file=1783%5croscube-x-bsp-qsg-l4t-32.5.0-kernel-1.0.8.pdf) in detail.

### Installing dependencies
Clone `tier4/perception_ecu_container` and move to the directory.

```bash
git clone https://github.com/tier4/perception_ecu_container.git
cd perception_ecu_container
```

You can install the dependencies using the provided Ansible script.

```bash
./setup-dev-env.sh
sudo reboot
```

### Building docker
```
./docker/build.sh
```

### Running docker
```
./docker/run.sh
```

This script mounts `perception_ecu_container` directory onto `/workspace` in docker container.

For example, the followings shows how to execute single camera object detection
```bash
./docker/run.sh
source /workspace/install/setup.bash  # Don't miss enabling workspace ROS packages
ros2 launch perception_ecu_or_launch perception_ecu_or.launch.xml
```

### TIER IV Automotive HDR Camera device driver

> - The contents of this repository are tested using [TIER IV Automotive HDR Camera C1](https://sensor.tier4.jp/automotive-hdr-camera). Device driver are provided at https://github.com/tier4/tier4_automotive_hdr_camera 

Reviewer: typo. It's good to check sentences using Grammarly or something else.

```suggestion
- The contents of this repository are tested using [TIER IV Automotive HDR Camera C1](https://sensor.tier4.jp/automotive-hdr-camera). Device drivers are provided at https://github.com/tier4/tier4_automotive_hdr_camera 
```

Author: https://github.com/tier4/perception_ecu_container/pull/1/commits/fe9a4ad2f69e2f2e76dc354da67043cd13319a2d

> - Please follow the installation guide [[English](https://github.com/tier4/tier4_automotive_hdr_camera/releases/download/v1.1.0/TIER.IV.Automotive.HDR.Camera.C1.Quick.Start.Guide.Nvidia.Jetson.AGX.Orin.DevKit._EN_v0_1_1.pdf)] [[Japanese](https://github.com/tier4/tier4_automotive_hdr_camera/releases/download/v1.1.0/TIER.IV.Automotive.HDR.Camera.C1.Quick.Start.Guide.Nvidia.Jetson.AGX.Orin.DevKit._JPN_v0_1_1.pdf)] to install the device driver manually.

Reviewer: [style] To follow Google's guidelines.
- https://developers.google.com/style/tone#politeness-and-use-of-please
- https://developers.google.com/style/clause-order

```suggestion
- To install the device driver manually, follow the installation guide [[English](https://github.com/tier4/tier4_automotive_hdr_camera/releases/download/v1.1.0/TIER.IV.Automotive.HDR.Camera.C1.Quick.Start.Guide.Nvidia.Jetson.AGX.Orin.DevKit._EN_v0_1_1.pdf)] [[Japanese](https://github.com/tier4/tier4_automotive_hdr_camera/releases/download/v1.1.0/TIER.IV.Automotive.HDR.Camera.C1.Quick.Start.Guide.Nvidia.Jetson.AGX.Orin.DevKit._JPN_v0_1_1.pdf)].
```

Author: https://github.com/tier4/perception_ecu_container/pull/1/commits/fe9a4ad2f69e2f2e76dc354da67043cd13319a2d

Author: https://github.com/tier4/perception_ecu_container/pull/1/commits/212dc6363e76e1566635877bbb44475ac9300a12

    - **Note: The latest device driver will be installed automatically during executing setup script in the following section. **
    - [For RQX-58G users] The camera type (i.e., C1/C2) that can connect is fixed for each GMSL2 port. The following table shows the default setting:

      | GMSL2 port No.          | 1  | 2  | 3  | 4  | 5  | 6  | 7  | 8  |
      |-------------------------|----|----|----|----|----|----|----|----|
      | connectable camera type | C1 | C1 | C2 | C2 | C1 | C1 | C2 | C2 |

## Repository overview
- [tier4/perception_ecu_container](https://github.com/tier4/perception_ecu_container)
    - Meta-repository containing `.repos` file to construct perception-ecu workspace
- [tier4/perception_ecu_launch](https://github.com/tier4/perception_ecu_launch.git)
- [tier4/sensor_trigger](https://github.com/tier4/sensor_trigger.git)
- [tier4/ros2_v4l2_camera](https://github.com/tier4/ros2_v4l2_camera.git)
- [tier4/perception_ecu_individual_params](https://github.com/tier4/perception_ecu_individual_params)

> - [autowarefoundation/autoware.universe](https://github.com/autowarefoundation/autoware.universe.git)

Reviewer: [style] It's better to apply Prettier. If you install `pre-commit.ci`, it's automatically done.

Author: https://github.com/tier4/perception_ecu_container/pull/1/commits/212dc6363e76e1566635877bbb44475ac9300a12

# Perception ECU Container

> This repository provides a sample ROS2 environment working on a Jetson AGX Xavier based ECU and GMSL2-compatible cameras.

Reviewer 2: ```suggestion
This repository provides a sample ROS2 environment for object detection using [tensorrt_yolox](https://github.com/autowarefoundation/autoware.universe/tree/main/perception/tensorrt_yolox). It works on a Jetson AGX Xavier-based ECU and GMSL2-compatible cameras.
```

Author: same?
https://github.com/tier4/perception_ecu_container/pull/1#discussion_r1058010357

> This repository provides a sample ROS2 environment working on a Jetson AGX Xavier based ECU and GMSL2-compatible cameras.

Reviewer 2: ```suggestion
This repository provides a sample ROS2 environment for object recognition using [tensorrt_yolox](https://github.com/autowarefoundation/autoware.universe/tree/main/perception/tensorrt_yolox). It works on a Jetson AGX Xavier-based ECU and GMSL2-compatible cameras.
```

Author: @kaz0227 Since we'll add more applications in the future, how about writing them arouind here instead of the top level of this readme?
https://github.com/tier4/perception_ecu_container/blob/bbda94b05318e598fd91c69b485adba27dea0b82/README.md?plain=1#L90

Reviewer 2: @aohsato I see. Let's add the comment there. Also, how about adding the comment here?
"As a sample application, the following images show object recognition results using the contents of this repository. We plan to add other applications here and release them as open-source software (OSS)."
https://github.com/tier4/perception_ecu_container/blob/1a93e75ce9f122f18d506383a5be2b999160c6c6/README.md?plain=1#L4

Author: @kaz0227 :cool: plz create your suggestion if you okay?

Reviewer 2: @aohsato Created a suggestion.
https://github.com/tier4/perception_ecu_container/pull/1/files#r1058147127

Author: @kaz0227 Thanks for your comment. Additionally, I added some tensorrt_yolox links, please review?
[1a2b288](https://github.com/tier4/perception_ecu_container/pull/1/commits/1a2b288afd6264366f8844343430ac041861817f)

> The following images show object recognition results using the contents of this repository.

Reviewer 2: ```suggestion
The following images show object detection results using the contents of this repository.
```

> The following images show object recognition results using the contents of this repository.

Reviewer 2: ```suggestion
As a sample application, the following images show object recognition results using the contents of this repository. We plan to add other OSS applications here.
```

![object recognition example](docs/sample.png "perception_ecu_container object recognition example")

> This repository is a docker-based ROS development environment. An overview of the system is shown below.

Reviewer 2: ```suggestion
This repository is a docker-based ROS2 environment. The system overview is shown below.
```

![system overview](docs/overview.drawio.svg "perception_ecu_container overview")

## Getting Started

### Prerequisites

- Camera: v4l2 compatible cameras, including [TIER IV Automotive HDR Camera C1](https://sensor.tier4.jp/automotive-hdr-camera)
- ECU: Jetson AGX Xavier based ECU, including [RQX-58G](https://www.adlinktech.com/Products/ROS2_Solution/ROS2_Controller/RQX-580_58G) from ADLINK Technology Inc. and [Developer Kit](https://www.nvidia.com/ja-jp/autonomous-machines/embedded-systems/jetson-agx-xavier) from NVIDIA Corp.
- Board support packages: NVIDIA L4T R32.6.1 (including Ubuntu 18.04) or higher

**NOTE: BSP installation for ADLINK RQX-58G**

> This repository assumes that RQX-58Gs has been properly configured by following the official quick start guide from ADLINK Technology, Inc. Please see the [official document](https://www.adlinktech.com/Products/Download.ashx?type=MDownload&isQuickStart=yes&file=1783%5croscube-x-bsp-qsg-l4t-32.5.0-kernel-1.0.8.pdf) in detail.

Reviewer 2: ```suggestion
This repository assumes that RQX-58Gs has been properly configured by following the official quick start guide from ADLINK Technology, Inc. Please see the [official document](https://www.adlinktech.com/Products/Download.ashx?type=MDownload&isQuickStart=yes&file=1783%5croscube-x-bsp-qsg-l4t-32.5.0-kernel-1.0.8.pdf) in detail. To download the BSP image, please visit the ADLINK official page [here](https://www.adlinktech.com/Products/DownloadSoftware.aspx?lang=en&pdNo=1783&MainCategory=ROS2-Solution.aspx&kind=BS). (If you are accessing the site for the first time, you will be prompted to create an account.)
```

TIER IV camera driver ([tier4/tier4_automotive_hdr_camera](https://github.com/tier4/tier4_automotive_hdr_camera)), is already installed in the RQX-58G BSP image and it can also be updated during your setup process.

An example connection between ECUs:

![system connection example](docs/connection.drawio.svg "system connection example")

### Installing dependencies

As a first step, clone `tier4/perception_ecu_container` and move to the directory.

```sh
git clone https://github.com/tier4/perception_ecu_container.git
cd perception_ecu_container
```

> You can install the dependencies using the provided ansible script. You will be asked if you want to install the TIER IV camera driver, press `N` button if you want to skip this step because it is pre-installed or for some other reason.

Reviewer 2: ```suggestion
You can install the dependencies using the provided ansible script. During the installation process, you will be asked if you want to install the TIER IV camera driver. If you already have the driver installed and want to skip this step, please type `N` to continue.
```

```sh
./setup-dev-env.sh

[Warning] Do you want to install/update the TIER IV camera driver? [y/N]:
```

> Finally, reboot once for dependencies and permission settings to take effect.

Reviewer 2: ```suggestion
Finally, please reboot the system to make the installed dependencies and permission settings effective.
```

```sh
sudo reboot
```

### Building docker

Build your docker image and build the ROS execution environment.

```sh
./docker/build.sh
```

### Building ROS workspace

Create your ROS workspace and clone repositories using vcstool.

```
mkdir src
vcs import src < perception_ecu.repos
```

If you want to update cloned repositories, use the following command.

```
vcs import src < perception_ecu.repos
vcs pull src
```

> Build your ROS workspace using your built docker image.

Reviewer 2: ```suggestion
Build your ROS workspace using your built docker image. This process would take about 10 minutes.
```
